### ISSUE TITLE
Error "cannot infer type" when using '?' in async block + bad diagnostic

### ISSUE THREAD

**Issue by russelltg** [rust-lang/rust#63502]
This seems related to #42424, except I don't see an obvious workaround like in #42424.

```rust
#![feature(async_await)]

use std::io::Error;

fn make_unit() -> Result<(), Error> { 
    Ok(())
}

fn main() {
    let fut = async {
        make_unit()?;
        
        Ok(())
    };
}
```

Fails with the error

```
error[E0282]: type annotations needed for `impl std::future::Future`
  --> src/main.rs:11:9
   |
10 |     let fut = async {
   |         --- consider giving `fut` the explicit type `impl std::future::Future`, with the type parameters specified
11 |         make_unit()?;
   |         ^^^^^^^^^^^^ cannot infer type
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ea7e5f7b6e1637f6a39aee0b8209c99e

First of all, it seems like this shouldn't be an error--it should be able to deduce that the return type is `Result<(), io::Error>`, but also the diagnostic **does not work**. If you try to add `impl std::future::Future<Output=Result<(), Error>>` as a type annotation, it also fails to compile because `impl Trait` is only allowed as return types and argument types.

**Comment by estebank** [comment]
Opened https://github.com/rust-lang/rust/issues/63504 for the incorrect suggestion. For the main issue, `?` interacts in non-obvious ways with type inference because it does implicit conversion. In this case, because `fut` doesn't explicitly constraint what the expected type is, `rustc` doesn't know what `Result<(), std::io:Error>` should be converted to.

There are other similar tickets filed about these kind of type inference issues: https://github.com/rust-lang/rust/issues/49391, https://github.com/rust-lang/rust/issues/38508, https://github.com/rust-lang/rust/issues/46333, https://github.com/rust-lang/rust/issues/46680, https://github.com/rust-lang/rust/issues/48089, https://github.com/rust-lang/rust/issues/61152, https://github.com/rust-lang/rust/issues/58517 and https://github.com/rust-lang/rust/issues/63082.

(larger work around `?` tracked in https://github.com/rust-lang/rust/issues/31436)

**Comment by cramertj** [comment]
> I don't see an obvious workaround

The workaround is this:
```rust
use std::io::Error;

fn make_unit() -> Result<(), Error> { 
    Ok(())
}

fn main() {
    let fut = async {
        make_unit()?;
        
        Ok::<(), Error>(())
    };
}
```

**Comment by estebank** [comment]
For this case, we probably could suggest a temporary local binding with an explicit type:

```
    let fut = async {
        let x: Result<_, Error> = make_unit();
        x?
    };
```

As for the syntax, we could borrow the closure syntax and allow something like `async -> Result<(), Error> {}`, but that is uuuugly.

**Comment by cramertj** [comment]
@estebank did you mean to include the `?` after the `x` in your example above?

**Comment by estebank** [comment]
Yes, but was missing the final expr `Ok(())` to be correctly equivalent the original code. Just updated it.

**Comment by cramertj** [comment]
@estebank That code still would not compile.

**Comment by estebank** [comment]
You're absolutely right. Need more coffee.

**Comment by programmerjake** [comment]
doesn't `?` still use `From` rather than `Into`, or did I miss some happy RFC?

**Comment by Dylan-DPC** [comment]
Current output:
```
error[[E0282]](https://doc.rust-lang.org/nightly/error_codes/E0282.html): type annotations needed
  --> src/main.rs:11:9
   |
11 |         Ok(())
   |         ^^ cannot infer type of the type parameter `E` declared on the enum `Result`
   |
help: consider specifying the generic arguments
   |
11 |         Ok::<(), E>(())
   |           +++++++++
```

**Comment by estebank** [comment]
The only thing left is for inference to detect that `E` could have been `Error` and suggest that.

**Comment by KalyanKadiyala** [comment]
Not a question, but a comment with something that I tried to resolve annoyance at call-site. The `?`-operator is used to extract value from a Result. Wanted to check, if there's been any progress on the topic (less async itself).  The [coherence](https://doc.rust-lang.org/reference/items/implementations.html#trait-implementation-coherence) rules prevent adding implicit coercion support for concrete types in my case. This can be the case with any other remotely defined Traits. 

Briefly,

- When I explicitly provide type annotations, the call-site assignment is inferred to expected error type on the Result return value.

<img width="701" alt="Image" src="https://github.com/user-attachments/assets/c2326c42-eb9b-49e0-82e6-3af93a3a8ce1" />

- However, if I leave it to inference, the valid inferred result loses type specificity on the error variant, as shown below:

<img width="697" alt="Image" src="https://github.com/user-attachments/assets/edaed8a7-7908-4560-80f0-180bd1dd67ae" />

Though a deviation from the standard approach of using the `?` operator, I've added an implementation support that wraps ? operator usage with expected concrete types on the Result type using a separate extract method. Here's the impl:

<img width="497" alt="Image" src="https://github.com/user-attachments/assets/f5afe7a0-0037-4021-b0aa-9a4c1b3a9508" />

And here's the call-site, where the assigned value from the closure gets resolved (by inference) to expected Result's error value type:

<img width="531" alt="Image" src="https://github.com/user-attachments/assets/c75125cf-2066-4186-8427-6dc142c52023" />

Like to share this and seek any comments/suggestions. I'm using Rust 1.86 stable build.
